service/dap/server.go

```diff
@@ -686,16 +689,222 @@ func (s *Server) onStackTraceRequest(request *dap.StackTraceRequest) {
+	if err != nil {
+		s.sendErrorResponse(request.Request, UnableToListLocals, "Unable to list local vars", err.Error())
+		return
+	}
+	locScope := api.Variable{Name: "Locals", Children: locals}
+
+	// TODO(polina): Annotate shadowed variables
+	// TODO(polina): Retrieve global variables
+
+	scopeArgs := dap.Scope{Name: argScope.Name, VariablesReference: s.variableHandles.create(argScope)}
+	scopeLocals := dap.Scope{Name: locScope.Name, VariablesReference: s.variableHandles.create(locScope)}
+	scopes := []dap.Scope{scopeArgs, scopeLocals}
+
+	response := &dap.ScopesResponse{
+		Response: *newResponse(request.Request),
+		Body:     dap.ScopesResponseBody{Scopes: scopes},
+	}
+	s.send(response)
 }
 
-// onVariablesRequest sends a not-yet-implemented error response.
+// onVariablesRequest handles 'variables' requests.
 // This is a mandatory request to support.
-func (s *Server) onVariablesRequest(request *dap.VariablesRequest) { // TODO V0
-	s.sendNotYetImplementedErrorResponse(request.Request)
+func (s *Server) onVariablesRequest(request *dap.VariablesRequest) {
+	variable, ok := s.variableHandles.get(request.Arguments.VariablesReference)
+	if !ok {
+		s.sendErrorResponse(request.Request, UnableToLookupVariable, "Unable to lookup variable", fmt.Sprintf("unknown reference %d", request.Arguments.VariablesReference))
+		return
+	}
+	v := variable.(api.Variable)
+	children := make([]dap.Variable, 0)
+	// TODO(polina): check and handle if variable loaded incompletely
+	// https://github.com/go-delve/delve/blob/master/Documentation/api/ClientHowto.md#looking-into-variables
+
+	switch v.Kind {
+	case reflect.Map:
+		kvIndex := 0
+		for i := range v.Children {
+			// A map will have twice as many children as there are elements.
+			// Process children in pairs: even indices are map keys, odd indices are values.
+			if i%2 == 0 {
+				kvIndex = i / 2
+				continue
+			}
+			key, keyref := s.convertVariable(v.Children[i-1])
+			val, valref := s.convertVariable(v.Children[i])
+			// If key or value or both are scalars, we can use
+			// a single variable to represet key:value format.
+			// Otherwise, we must return separate variables for both.
+			if keyref > 0 && valref > 0 { // Both are not scalars
+				keyvar := dap.Variable{
+					Name:               fmt.Sprintf("[key %d]", kvIndex),
+					Value:              key,
+					VariablesReference: keyref,
+				}
+				valvar := dap.Variable{
+					Name:               fmt.Sprintf("[val %d]", kvIndex),
+					Value:              val,
+					VariablesReference: valref,
+				}
+				children = append(children, keyvar, valvar)
+			} else { // At least one is a scalar
+				kvvar := dap.Variable{
+					Name:               key,
+					Value:              val,
+					VariablesReference: valref,
+				}
+				if keyref != 0 {
+					kvvar.VariablesReference = keyref
+				} else if valref != 0 {
+					kvvar.VariablesReference = valref
+				}
+				children = append(children, kvvar)
```

Comment:
I was worried about this, but decided that since we only expand the key while the value is obviously a non-expandable scalar, then it will not be confusing. The problem with the original TypeScript implementation is that it does not support key expansion in any scenario. It only supports non-scalar values. So instead of feature parity, I decided to actually fix things here. I am not super happy with how things look, but can't think of a better way. I am open to suggestions.

Before:
![image](https://user-images.githubusercontent.com/51177946/89358181-8344d500-d677-11ea-81f8-3c03c639c234.png)
After:
![image](https://user-images.githubusercontent.com/51177946/89358342-eb93b680-d677-11ea-9a99-7beff99d6284.png)

Reply:
It looks somewhat strange but better than I thought, I guess it's fine.

Reply:
I found a bug and just pushed a fix for this case. Even though I was returning the right number of elements, vscode only displayed a single one because all compound keys had the same value of their type. Now this looks like so:
![image](https://user-images.githubusercontent.com/51177946/89531205-b769e400-d7a4-11ea-9050-db638414a7f9.png)